src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -161,4 +176,124 @@ ErrorMessageOr<uint64_t> AllocateMemoryForTrampolines(pid_t pid, const AddressRa
+    ErrorMessageOr<int32_t> new_displacement_or_error =
+        AddressDifferenceAsInt32(old_absolute_address, new_address + instruction->size);
+    if (new_displacement_or_error.has_error()) {
+      return ErrorMessage(absl::StrFormat(
+          "While trying to relocate an instruction with rip relative addressing the target was out "
+          "of range from the trampoline. old address: %#x, new address :%#x instruction: %s",
+          old_address, new_address, InstructionBytesAsString(instruction)));
+    }
+    result.code.resize(instruction->size);
+    memcpy(result.code.data(), instruction->bytes, instruction->size);
+    *absl::bit_cast<int32_t*>(result.code.data() + instruction->detail->x86.encoding.disp_offset) =
+        new_displacement_or_error.value();
+  } else if (instruction->detail->x86.opcode[0] == 0xeb ||
+             instruction->detail->x86.opcode[0] == 0xe9) {
+    // Unconditional jump to relative immediate parameter (32 bit or 8 bit).
+    // We compute the absolute address and jump there:
+    // jmp [rip + 0]                ff 25 00 00 00 00
+    // .byte absolute_address       01 02 03 04 05 06 07 08
+    const int32_t immediate =
+        (instruction->detail->x86.opcode[0] == 0xe9)
+            ? *absl::bit_cast<int32_t*>(instruction->bytes +
+                                        instruction->detail->x86.encoding.imm_offset)
+            : *absl::bit_cast<int8_t*>(instruction->bytes +
+                                       instruction->detail->x86.encoding.imm_offset);
+    const uint64_t absolute_address = old_address + instruction->size + immediate;
+    MachineCode code;
+    code.AppendBytes({0xff, 0x25}).AppendImmediate32(0x0000000).AppendImmediate64(absolute_address);
+    result.code = code.GetResultAsVector();
+    result.position_of_absolute_address = 6;
+  } else if (instruction->detail->x86.opcode[0] == 0xe8) {
+    // Call function at relative immediate parameter.
+    // We compute the absolute address of the called function and call it like this:
+    // Call [rip+2]                 ff 15 02 00 00 00
+    // jmp label;                   eb 08
+    // .byte absolute_address       01 02 03 04 05 06 07 08
+    // label:
+    const int32_t immediate = *absl::bit_cast<int32_t*>(
+        instruction->bytes + instruction->detail->x86.encoding.imm_offset);
+    const uint64_t absolute_address = old_address + instruction->size + immediate;
+    MachineCode code;
+    code.AppendBytes({0xff, 0x15})
+        .AppendImmediate32(0x00000002)
+        .AppendBytes({0xeb, 0x08})
+        .AppendImmediate64(absolute_address);
+    result.code = code.GetResultAsVector();
+    result.position_of_absolute_address = 8;
+  } else if ((instruction->detail->x86.opcode[0] & 0xf0) == 0x70) {
+    // 0x7? are conditional jumps to an 8 bit immediate.
+    // We invert the condition of the jump and construct the following code sequence.
+    // j(!cc) 0x0e                  7? 0e
```

Comment:
So 0x0e makes you jump after the absolute_address, yes?

Reply:
exactly, added a comment.